services/transientsource/transientsource.go

```diff
@@ -0,0 +1,140 @@
+package transientsource
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	backendconfig "github.com/rudderlabs/rudder-server/config/backend-config"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/utils/pubsub"
+	"github.com/tidwall/gjson"
+)
+
+// Service provides services related to transient source ids
+type Service interface {
+
+	// SourceIdsSupplier provides an up-to-date supplier of transient source ids
+	SourceIdsSupplier() func() []string
+
+	// Apply performs transient source_id filtering logic against a single source_id.
+	// If it corresponds to a transient source_id, the result will be true, otherwise false.
+	Apply(sourceId string) bool
+
+	// ApplyJob performs transient source_id filtering logic against a single job.
+	// If the job corresponds to a transient source_id, the result will be true, otherwise false.
+	ApplyJob(job *jobsdb.JobT) bool
+
+	// ApplyParams performs transient source_id filtering logic against a job's parameters.
+	// If the parameters contain a transient source_id, the result will be true, otherwise false.
+	ApplyParams(params json.RawMessage) bool
+}
+
+// NewService creates a new service that updates its transient source ids while
+// backend configuration gets updated.
+func NewService(ctx context.Context, config backendconfig.BackendConfig) Service {
+	s := &service{
+		init: make(chan struct{}),
+	}
+	s.updateLoop(ctx, config)
+	return s
+}
+
+// NewEmptyService creates a new service that operates against an empty list of transient source ids
+// Useful for tests, when you are not interested in testing for transient sources.
+func NewEmptyService() Service {
+	return NewStaticService([]string{})
+}
+
+// NewStaticService creates a new service that operates against a predefined list of transient source ids.
+// Useful for tests.
+func NewStaticService(sourceIds []string) Service {
+	s := &service{
+		init:         make(chan struct{}),
+		sourceIds:    sourceIds,
+		sourceIdsMap: asMap(sourceIds),
+	}
+	close(s.init)
+	return s
+}
+
+type service struct {
+	init         chan struct{}
+	sourceIds    []string
+	sourceIdsMap map[string]struct{}
+}
+
+func (r *service) SourceIdsSupplier() func() []string {
+	return func() []string {
+		<-r.init
+		return r.sourceIds
+	}
+}
+
+func (r *service) Apply(sourceId string) bool {
+	<-r.init
+	_, ok := r.sourceIdsMap[sourceId]
+	return ok
+}
+
+func (r *service) ApplyParams(params json.RawMessage) bool {
+	sourceId := gjson.GetBytes(params, "source_id").String()
+	return r.Apply(sourceId)
+}
+
+func (r *service) ApplyJob(job *jobsdb.JobT) bool {
+	return r.ApplyParams(job.Parameters)
+}
+
+// updateLoop uses backend config to retrieve & keep up-to-date the list of transient source ids
+func (r *service) updateLoop(ctx context.Context, config backendconfig.BackendConfig) {
+
+	var onceInit sync.Once
+	go func() {
```

Comment:
Minor but still worth mentioning. This kind of APIs where a function seems synchronous but is not should be avoided.

So instead of:

```go
func someFunctionThatSeemsSynchronous() {
    go func() {
        // let's do stuff asynchronously instead
    }()
}
```

We should consider doing:

```go
func someFunctionThatIsActuallySynchronous() {
    // let's do stuff synchronouly!!!
}

func someCaller() {
    // if I need to call it synchronously then I do
    someFunctionThatIsActuallySynchronous()

    // but if I need to go asynchronous then I do
    go someFunctionThatIsActuallySynchronous()
}
```

The advantages are:
1. the name of the function is not misleading anymore
2. it's up to the caller whether it wants to go sync or async
3. the caller might want to call multiple functions asynchronouly so having the way I'm proposing would simplify the go routines coordination, which should be on the caller side

So I suggest you do ` go updateLoop()` and remove the go routine from inside it.

As a sidenote `onceInit` could potentially be part of the `struct` and do `r.onceInit.Do(...)`.

Reply:
Thanks, you are correct!